src/dtx/dtx_common.c

```diff
@@ -0,0 +1,592 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * GOVERNMENT LICENSE RIGHTS-OPEN SOURCE SOFTWARE
+ * The Government's rights to use, modify, reproduce, release, perform, display,
+ * or disclose this software are subject to the terms of the Apache License as
+ * provided in Contract No. B609815.
+ * Any reproduction of computer software, computer software documentation, or
+ * portions thereof marked with this legend must also reproduce the markings.
+ */
+/**
+ * dtx: DTX common logic
+ */
+#define D_LOGFAC	DD_FAC(dtx)
+
+#include <abt.h>
+#include <uuid/uuid.h>
+#include <daos_srv/pool.h>
+#include <daos_srv/container.h>
+#include <daos_srv/vos.h>
+#include <daos_srv/dtx_srv.h>
+#include <daos_srv/daos_server.h>
+#include "dtx_internal.h"
+
+struct dtx_batched_commit_args {
+	d_list_t		 dbca_link;
+	struct ds_cont_hdl	*dbca_cont_hdl;
+};
+
+void
+dtx_aggregate(void *arg)
+{
+	struct ds_cont		*cont = arg;
+
+	while (!cont->sc_closing) {
+		int	rc;
+
+		rc = vos_dtx_aggregate(cont->sc_hdl, DTX_AGG_YIELD_INTERVAL,
+				       DTX_AGG_THRESHOLD_AGE_LOWER);
+		if (rc != 0)
+			break;
+
+		ABT_thread_yield();
+	}
+
+	cont->sc_dtx_aggregating = 0;
+	ds_cont_put(cont);
+}
+
+static void
+dtx_flush_committable(struct ds_cont_hdl *cont_hdl)
+{
+	struct ds_cont		*cont = cont_hdl->sch_cont;
+	ABT_future		 future;
+	int			 rc;
+
+	do {
+		struct dtx_entry	*dte = NULL;
+
+		rc = vos_dtx_list_committable(cont->sc_hdl,
+					      DTX_THRESHOLD_COUNT, &dte);
+		if (rc <= 0)
+			break;
+
+		rc = dtx_commit(cont_hdl->sch_pool->spc_uuid, cont->sc_uuid,
+				dte, rc, cont_hdl->sch_pool->spc_map_version);
+		D_FREE(dte);
+	} while (rc >= 0);
+
+	if (rc < 0)
+		D_ERROR(DF_UUID": Fail to flush CoS cache: rc = %d\n",
+			DP_UUID(cont->sc_uuid), rc);
+
+	D_ASSERT(cont->sc_dtx_flush_cbdata != NULL);
+
+	future = cont->sc_dtx_flush_cbdata;
+	rc = ABT_future_set(future, NULL);
+	D_ASSERTF(rc == ABT_SUCCESS, "ABT_future_set failed for DTX flush on "
+		  DF_UUID": rc = %d\n", DP_UUID(cont->sc_uuid), rc);
+}
+
+void
+dtx_batched_commit(void *arg)
+{
+	struct dss_module_info	*dmi = dss_get_module_info();
+
+	while (1) {
+		ABT_bool			 state;
+		struct dtx_batched_commit_args	*dbca;
+		struct ds_cont_hdl		*cont_hdl;
+		struct ds_cont			*cont;
+		struct dtx_entry		*dte = NULL;
+		struct dtx_stat			 stat = { 0 };
+		int				 rc;
+
+		if (d_list_empty(&dmi->dmi_dtx_batched_list))
+			goto check;
+
+		dbca = d_list_entry(dmi->dmi_dtx_batched_list.next,
+				    struct dtx_batched_commit_args, dbca_link);
+		cont_hdl = dbca->dbca_cont_hdl;
+		cont = cont_hdl->sch_cont;
+		if (cont->sc_closing) {
+			dtx_flush_committable(cont_hdl);
+			/* XXX: There may be resent-close during DTX flush,
```

Comment:
This is another example that illustrates the subtle point above though a little nit-picky.   "The close may have been resent during DTX flush".   Or, if you want to use an adjective, "There may be a duplicate close during DTX flush".

Reply:
I will update the comment.